browser/importer/brave_profile_writer.cc

```diff
@@ -63,3 +79,199 @@ void BraveProfileWriter::UpdateStats(const BraveStats& stats) {
+  bridge_ptr_ = bridge;
+}
+
+void BraveProfileWriter::OnWalletInitialized(brave_rewards::RewardsService*
+  rewards_service, int error_code) {
+  if (error_code) {
+    // Cancel the import if wallet creation failed
+    std::ostringstream msg;
+    msg << "An error occurred while trying to create a "
+      << "wallet to restore into (error_code=" << error_code << ")";
+    CancelWalletImport(msg.str());
+    return;
+  }
+
+  LOG(INFO) << "Wallet creation successful\nStarting wallet recovery...";
+  rewards_service->RecoverWallet(ledger_.passphrase);
+}
+
+void BraveProfileWriter::BackupWallet() {
+  const base::FilePath profile_default_directory = profile_->GetPath();
+  std::ostringstream backup_filename;
+  backup_filename << "ledger_import_backup_"
+    << base::NumberToString((unsigned long long)base::Time::Now().ToJsTime());
+
+  LOG(INFO) << "Making backup of current \"ledger_state\" as "
+    << "\"" << backup_filename.str() << "\"";
+
+  base::PostTaskAndReplyWithResult<bool, bool>(
+    task_runner_.get(),
+    FROM_HERE,
+    base::Bind(&base::CopyFile,
+      profile_default_directory.AppendASCII("ledger_state"),
+      profile_default_directory.AppendASCII(backup_filename.str())),
+    base::Bind(&BraveProfileWriter::OnWalletBackupComplete,
+      this));
+}
+
+void BraveProfileWriter::OnWalletBackupComplete(bool result) {
+  if (!result) {
+    CancelWalletImport("Failed to make a backup of \"ledger_state\"");
+    return;
+  }
+
+  LOG(INFO) << "Backup complete; Recovering imported wallet...";
+  rewards_service_->RecoverWallet(ledger_.passphrase);
+}
+
+void BraveProfileWriter::OnWalletProperties(
+  brave_rewards::RewardsService* rewards_service,
+  int error_code,
+  brave_rewards::WalletProperties* properties) {
+  // Avoid overwriting Brave Rewards wallet if:
+  // - it existed BEFORE import happened
+  // - it has a non-zero balance
+  if (properties->balance > 0) {
+    std::ostringstream msg;
+    msg << "Brave Rewards wallet existed before import and "
+      << "has a balance of " << properties->balance << "; skipping "
+      << "Brave Payments import.";
+    CancelWalletImport(msg.str());
+    return;
+  } else {
+    LOG(INFO) << "Existing wallet does not have a balance";
+  }
+
+  BackupWallet();
+}
+
+void BraveProfileWriter::OnRecoverWallet(
+    brave_rewards::RewardsService* rewards_service,
+    unsigned int result,
+    double balance,
+    std::vector<brave_rewards::Grant> grants) {
+  rewards_service->RemoveObserver(this);
+
+  if (result) {
+    // Cancel the import if wallet restore failed
+    std::ostringstream msg;
+    msg << "An error occurred while trying to restore "
+      << "the wallet (result=" << result << ")";
+    CancelWalletImport(msg.str());
+    return;
+  }
+
+  LOG(INFO) << "Wallet restore successful";
+  SetWalletProperties(rewards_service);
+
+  // Set the pinned item count (rewards can detect and take action)
+  PrefService* prefs = profile_->GetOriginalProfile()->GetPrefs();
```

Comment:
why using  `GetOriginalProfile` here? You can use profile_ directly because we won't do import on incognito profile.